### DISCUSSION TITLE
dataBank: zef_dataBank_setData causes a memory leak and print issues of some functions

### DISCUSSION THREAD

**Discussion by JoonasJL** [sampsapursiainen/zeffiro_interface#173] [Ideas]
Function `zef_dataBank_setData` has the following if-chain leading it to call itself infinitely, or a bit less
``` matlab
if zef.dataBank.loadParents
    
        zef.dataBank.hash=reverse(zef.dataBank.hash);
        zef.dataBank.hash=extractAfter(zef.dataBank.hash, '_');
        zef.dataBank.hash=reverse(zef.dataBank.hash);

        if ~strcmp(zef.dataBank.hash, 'node') 
            **zef_dataBank_setData;**
        end   
```

Now that the `zef` structure is taken in the function, this will cause out of memory error.

It seems also that some of the functions are missing a semicolon in the starting lines
```matlab
if nargin == 0
zef = evalin('base','zef')
end
```

Personally, I could make some edits to make this work more efficiently because I have done a lot of work with fields and became good at it, but on the other hand my time is currently limited and the style of scripting is so far from mine that it will take a while for me to understand what there is happening under the hood.

**Comment by SeSodesa** [comment]
If `zef` was written as a [handle class] instead of being a `struct`, the copy-on-write semantics of value classes like `struct`s would not be applied to it. Since `structs` are value classes, when one does

    zef2 = zef;

    zef2.field = value;

the value behind `zef2` gets copied in memory, so that `zef` and `zef2` no longer point at the same instance. If `zef` was a handle class, this would not happen, and both `zef` and `zef2` would keep pointing to the same value on the heap or dynamic memory, which is where Matlab stores everything except variable symbols or names, much like Python does.

Why is this important? Because when you pass a name such as `zef` into a function in Matlab, Matlab essentially treats it like `zef2` was treated above, as a reference to an existing value but with a differently scoped name. Whether the value gets copied when/if the function assigns something to a field of the input depends entirely on whether the input argument is a value or a handle class.

Here is a comparison between value- and handle classes: <https://se.mathworks.com/help/matlab/matlab_oop/comparing-handle-and-value-classes.html>.

[handle class]: https://se.mathworks.com/help/matlab/handle-classes.html

**Reply by SeSodesa** [reply]
There were some tricks around avoiding the unnecessary copying of value- or copy-on-write classes, but these are not always very reliable, since functions have to be written in a very specific way for it to work: <https://se.mathworks.com/help/matlab/matlab_prog/avoid-unnecessary-copies-of-data.html>.

**Reply by SeSodesa** [reply]
Handle classes can be copied, if instead of inheriting from the `handle` class directly, you inherit from its subclass, `matlab.mixin.Copyable`:

    >> superclasses("matlab.mixin.Copyable")

    Superclasses for class matlab.mixin.Copyable:

        handle

Once you do `MyHandleClass < matlab.mixin.Copyable`, you can rewrite the `copy` and `copyElement` methods provided by the superclass to customize the copying behaviour: <https://se.mathworks.com/help/matlab/matlab_oop/custom-copy-behavior.html>. You then call these methods explicitly, when you want a copy of the value behind variable pointing to a class instance.
